routing/rt/visualizer/visualizer.go

```diff
@@ -0,0 +1,159 @@
+	HighlightedEdges []graph.Edge
+	Config           *VisualizerConfig
+	ApplyToNode      func(string) string
+	ApplyToEdge      func(*graph.ChannelInfo) string
+	pt               prefix_tree.PrefixTree
+	graphviz         *gographviz.Graph
+}
+
+func New(g *graph.Graph, highlightedNodes []graph.ID,
+	highlightedEdges []graph.Edge, config *VisualizerConfig) *visualizer {
+	if config == nil {
+		config = &DefaultVisualizerConfig
+	}
+	return &visualizer{
+		G:                g,
+		HighlightedNodes: highlightedNodes,
+		HighlightedEdges: highlightedEdges,
+		Config:           config,
+		ApplyToNode:      func(s string) string { return s },
+		ApplyToEdge:      func(info *graph.ChannelInfo) string { return "nil" },
+		pt:               prefix_tree.NewPrefixTree(),
+		graphviz:         gographviz.NewGraph(),
+	}
+}
+
+func (viz *visualizer) EnableShortcut(value bool) {
+	viz.Config.EnableShortcut = value
+}
+
+func (viz *visualizer) BuildPrefixTree() {
+	for _, node := range viz.G.GetIDs() {
+		id := node.String()
+		id = viz.ApplyToNode(id)
+		viz.pt.Add(id)
+	}
+}
+
+func (viz *visualizer) Draw() string {
+	viz.Base()
+	viz.DrawNodes()
+	viz.DrawEdges()
+	return viz.graphviz.String()
+}
+
+func (viz *visualizer) Base() {
+	viz.graphviz.SetName(viz.Config.GlobalCfg.Name)
+	viz.graphviz.SetDir(viz.Config.GlobalCfg.Dir)
+	viz.graphviz.SetStrict(viz.Config.GlobalCfg.Strict)
+	// TODO(evg): use viz.Add(...) instead viz.Attrs.Add(...)
+	viz.graphviz.Attrs.Add("bgcolor", viz.Config.GlobalCfg.BgColor)
+}
+
+func (viz *visualizer) DrawNodes() {
+	for _, node := range viz.G.GetIDs() {
+		if viz.IsHighlightedNode(node) {
+			viz.DrawNode(node, viz.Config.HighlightedNodeCfg)
+		} else {
+			viz.DrawNode(node, viz.Config.NodeCfg)
+		}
+	}
+}
+
+func (viz *visualizer) DrawNode(node graph.ID, cfg *NodeConfig) {
+	id := node.String()
+	id = viz.ApplyToNode(id)
+	if viz.Config.EnableShortcut {
+		// TODO(evg): processing errors
+		id, _ = viz.pt.Shortcut(id)
+	}
+	attrs := gographviz.Attrs{
+		"shape":     cfg.Shape,
+		"style":     cfg.Style,
+		"fontsize":  cfg.FontSize,
+		"fontcolor": cfg.FontColor,
+		"color":     cfg.Color,
+		"fillcolor": cfg.FillColor,
+	}
+	viz.graphviz.AddNode(viz.Config.GlobalCfg.Name, id, attrs)
+}
+
+func (viz *visualizer) IsHighlightedNode(node graph.ID) bool {
+	for _, value := range viz.HighlightedNodes {
+		if node.String() == value.String() {
+			return true
+		}
+	}
+	return false
+}
+
+func (viz *visualizer) DrawEdges() {
+	for _, edge := range viz.G.GetUndirectedEdges() {
+		if viz.IsHighlightedEdge(edge) {
+			viz.DrawEdge(edge, viz.Config.HighlightedEdgeCfg)
+		} else {
+			viz.DrawEdge(edge, viz.Config.EdgeCfg)
+		}
+	}
+}
+
+func (viz *visualizer) DrawEdge(edge graph.Edge, cfg *EdgeConfig) {
+	src := edge.Source().String()
+	src = viz.ApplyToNode(src)
+	tgt := edge.Target().String()
+	tgt = viz.ApplyToNode(tgt)
+	if viz.Config.EnableShortcut {
+		// TODO(evg): processing errors
+		src, _ = viz.pt.Shortcut(src)
```

Comment:
The errors here shouldn't be ignored. Also I think many of the functions should not be publicly exported. Instead, only the necessary public entrance points need to be exposed.